contrib/babelfishpg_tsql/sql/babelfishpg_tsql.sql

```diff
@@ -2973,71 +2970,149 @@ BEGIN
+					WITH parseTable AS (
+						SELECT (ROW_NUMBER() OVER (ORDER BY NULL)) as row,*
+						FROM STRING_SPLIT(@objname, '.'))
+					SELECT @subname = value FROM parseTable WHERE row = 3;
+				END
+			IF @name_count = 2
+				BEGIN
+					SET @schemaname = sys.schema_name();
+					WITH parseTable AS (
+						SELECT (ROW_NUMBER() OVER (ORDER BY NULL)) as row,*
+						FROM STRING_SPLIT(@objname, '.'))
+					SELECT @curr_relname = value FROM parseTable WHERE row = 1;
+					WITH parseTable AS (
+						SELECT (ROW_NUMBER() OVER (ORDER BY NULL)) as row,*
+						FROM STRING_SPLIT(@objname, '.'))
+					SELECT @subname = value FROM parseTable WHERE row = 2;
+				END
+			IF @name_count = 1
+				BEGIN
+					THROW 33557097, N'Either the parameter @objname is ambiguous or the claimed @objtype (COLUMN) is wrong.', 1;
+				END
+			
+			DECLARE @col_count INT;
+			SELECT @col_count = COUNT(*)FROM INFORMATION_SCHEMA.COLUMNS WHERE TABLE_NAME = @curr_relname and COLUMN_NAME = @subname;
+			IF @col_count < 0
+				BEGIN
+					THROW 33557097, N'There is no object with the given @objname.', 1;
+				END
+			SET @currtype = 'CO';
+			EXEC sys.babelfish_sp_rename_internal @subname, @newname, @schemaname, @currtype, @curr_relname;
+			PRINT 'Caution: Changing any part of an object name could break scripts and stored procedures.';
 		END
-	
-	DECLARE @count INT;
-	DECLARE @currtype char(2);
-	SELECT @count = COUNT(*) FROM sys.objects o1 INNER JOIN sys.schemas s1 ON o1.schema_id = s1.schema_id 
-	WHERE s1.name = @schemaname AND o1.name = @subname;
-	IF @count > 1
+	ELSE IF @objtype = 'OBJECT'
 		BEGIN
-			THROW 33557097, N'There are multiple objects with the given @objname.', 1;
+				IF @name_count > 3
+				BEGIN
+					THROW 33557097, N'No item by the given @objname could be found in the current database', 1;
+				END
+			IF @name_count = 3
+				BEGIN
+					WITH parseTable AS (
+						SELECT (ROW_NUMBER() OVER (ORDER BY NULL)) as row,*
+						FROM STRING_SPLIT(@objname, '.'))
+					SELECT @dbname = value FROM parseTable WHERE row = 1;
+					IF @dbname != sys.db_name()
+						BEGIN
+							THROW 33557097, N'No item by the given @objname could be found in the current database', 1;
+						END
+					WITH parseTable AS (
+						SELECT (ROW_NUMBER() OVER (ORDER BY NULL)) as row,*
+						FROM STRING_SPLIT(@objname, '.'))
+					SELECT @schemaname = value FROM parseTable WHERE row = 2;
+					WITH parseTable AS (
+						SELECT (ROW_NUMBER() OVER (ORDER BY NULL)) as row,*
+						FROM STRING_SPLIT(@objname, '.'))
+					SELECT @subname = value FROM parseTable WHERE row = 3;
+				END
+			IF @name_count = 2
+				BEGIN
+					WITH parseTable AS (
+						SELECT (ROW_NUMBER() OVER (ORDER BY NULL)) as row,*
+						FROM STRING_SPLIT(@objname, '.'))
+					SELECT @schemaname = value FROM parseTable WHERE row = 1;
+					WITH parseTable AS (
+						SELECT (ROW_NUMBER() OVER (ORDER BY NULL)) as row,*
+						FROM STRING_SPLIT(@objname, '.'))
+					SELECT @subname = value FROM parseTable WHERE row = 2;
```

Comment:
Let's create a helper proc or func to do the string split recursively. This sp_rename procedure block is too long and has too many duplications.